Ethereum-Max/Ethereum-Max-Contracts
# burn & reflect

# End to End Test Plan:
1. Deploy V1 EMax Coin
2. Verify Proxy owner - by default burn address
3. TransferOwnership of ProxyAdmin to Deployer
4. Add `setReflectRate`, `setBurnRate`, `burnPercentageOfTransaction` functions
5. Perform upgrade of implentation from `EmaxCoin` to `EmaxCoinV2`
6. Execute Setters of `BurnRate=3`, `ReflectionRate=3`

## Status:
* For our own testing we did an end to end of deploying a V1 proxy and implementation and then running an upgrade which was successful; However, we were not able to test the setters and getters due to the issues outlined below. In a separate deployment, when the reflection and burn logic is initalized the burn and reflect functionality appears to work properly.

## Deployment:
* Proxy
  * [`https://ropsten.etherscan.io/address/0x6dfad17f1da70ff9f579d30bc0881963b0c545ba#code`](https://ropsten.etherscan.io/address/0x6dfad17f1da70ff9f579d30bc0881963b0c545ba#code)
* EMaxCoinV2
  * [`https://ropsten.etherscan.io/address/0x9139ef4f2b8885f1c800892514bdc7d8df445572#code`](https://ropsten.etherscan.io/address/0x9139ef4f2b8885f1c800892514bdc7d8df445572#code)
* Implementation contract (according to Etherscan; more on this below)
  * [https://ropsten.etherscan.io/address/0x87e0b66c3b6c3c43cc8763d849015274faa4b9ae#code](https://ropsten.etherscan.io/address/0x87e0b66c3b6c3c43cc8763d849015274faa4b9ae#code)

* Logs showed `prepareUpgrade(V1 => V2)` was successful 


## Outstanding Tasks
* Last step we need to perform: EtherScan verification of the implementation contract. We noticed that the EMax smart contract deployed in mainnet has `read/write` proxy functions which we do not see on our test deployments nor a deployment of the original source code.

* [The transparent proxy](https://ropsten.etherscan.io/address/0x6dfad17f1da70ff9f579d30bc0881963b0c545ba#code) has 7 verified files listed under the contract tab. However [the main net Emax has 13](https://etherscan.io/token/0x15874d65e649880c2614e7a480cb7c9a55787ff6#readContract)

* We are pretty certain that this is due to us not verifying the implementation contract. We plan to look for help verifying this from EMax team as we havent been able to solve it (_errors from EtherScan compiler, more on this below_).

* We installed `truffle-verify` plugin. It successfully verified the Proxy & [EMaxCoinV2](https://ropsten.etherscan.io/address/0x9139ef4f2b8885f1c800892514bdc7d8df445572#code) links that were provided above. However, clicking the Proxy link above and clicking the more options button outlined in ScreenShot 1 shows an `is this a proxy` button, this leads to a messsage that our `implemenetation contract is not verified`. It links to [here](https://ropsten.etherscan.io/address/0x87e0b66c3b6c3c43cc8763d849015274faa4b9ae#code). We are fairly confident that successfully verifying the implementation contract will add the remaining write/read as proxy options that are visible on the main net proxy page.

* Final step: We successfully verified the burn and reflection functions worked when initializing them, we werent able to test the setters and getters because of the issue mentioned above. 

* Solution: Joint effort to get contract verified on EtherScan, we attempted multiple times to verify the contract on etherscan both manually and through the truffle verify plugin, while successful, the truffle verify plugin did not verify all contracts in the deployment, most importantly the implementation contract.

* Error from Etherscan: `bytecode mismatch error`
    * We are not certain what part of the code needs to be verified in the [implementation contract](https://ropsten.etherscan.io/address/0x87e0b66c3b6c3c43cc8763d849015274faa4b9ae#code).
